pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/instanceselector/InstanceSelector.java

```diff
@@ -0,0 +1,56 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.v2.instanceselector;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.model.ExternalView;
+import org.apache.pinot.common.request.BrokerRequest;
+
+
+/**
+ * The instance selector selects server instances to serve the query based on the selected segments.
+ */
+public interface InstanceSelector {
+
+  /**
+   * Initializes the instance selector with the enabled instances, external view and online segments (segments with
+   * ONLINE/CONSUMING instances in ideal state). Should be called only once before calling other methods.
+   */
+  void init(Set<String> enabledInstances, ExternalView externalView, Set<String> onlineSegments);
+
+  /**
+   * Processes the instances change. Changed instances are pre-computed based on the current and previous enabled
+   * instances only once on the caller side and passed to all the instance selectors.
+   */
+  void onInstancesChange(Set<String> enabledInstances, List<String> changedInstances);
+
+  /**
+   * Processes the external view change based on the given online segments (segments with ONLINE/CONSUMING instances in
+   * ideal state).
+   */
+  void onExternalViewChange(ExternalView externalView, Set<String> onlineSegments);
```

Comment:
Same here.. I think it will be great to explain the relevance of onlineSegments for future readers. We should indicate how they are used and when will they be empty/null

Reply:
It won't be null if it is not annotated as nullable. We don't use it right now, but maybe useful for future implementations. Added some description